### PR TITLE
authz: add audit logging APIs

### PR DESCRIPTION
**Pull request by rockspore**
Implements interfaces for pending gRFC A59: https://github.com/grpc/proposal/pull/346/files

RELEASE NOTES: N/A

### CODE REVIEW

**Comment by rockspore** [comment]
One small thing to note is that I removed `context.Context` from the `Log()` method of the logger. It's not really needed and to be consistent with other languages, we will not offer the method any additional information than the `AuditInfo`.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:60]: This means an `AuditLoggerBuilder` itself is also an `AuditLoggerConfig`, because it has a method with this same signature.  Maybe you wanted an unexported function in here, and implementers of `ParseAuditLoggerConfig` need to embed this interface in the type they implement?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:66]: If this is supposed to be allowed to block, I believe we do want a context here.  Otherwise, there is no way to cancel it or let it know what deadline it has.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:75]: Can this not error?  What if the logger is supposed to write to a file but the directory doesn't exist, for example?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:89]: `return m[name]` (which is `nil` if missing).

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:63]: What is an audit logger anyway?  How do you use one?  When would I use one?

Can you link to a doc or something?  This basically says nothing and is not very helpful for any user looking at the godoc.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:69]: Similar.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:43]: What is the format of this?  E.g. is this "service/method" or just the last part?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:3]: 2023

**Review by dfawley — CHANGES_REQUESTED** [review]

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:66]: Thanks for the review, Doug! I will work on the other comments but I'd like to discuss this particular comment a bit more. So `Log()` will be called during the authorization process, and I think we agreed that the same function in C++ and Java will only take the equivalent of this `AuditInfo`. So I believe there is no way there as well for the function to cancel or know anything about the deadline? If this is true, I feel we should keep it consistent here, although I do like passing a context to enable cancellation, which was why I had included it.

Comment (dfawley) [reply authz/audit_logger.go:76]: This seems like this is going to be a language difference to me, as the threading models are different in the different languages.  All blocking functions in Go should ideally accept contexts, but no similar constraint exists in Java/C++.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Added clarification + examples. It will be "/pkg.Service/Method"

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:62]: Good idea! I added an `auditLoggerConfig()` method and made it clear that users need to embed `AuditLoggerConfig` to implement their own config.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: This makes sense. I noticed methods like `IsAuthorized()` in this pkg also takes a context whereas I don't believe C++/Java has something similar. I added the context back.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:97]: If `Build()` returns an error at this stage, the logger will be nil but the data plane will have no way to NACK the filter. I think the invalidation you mentioned above should be done by `ParseAuditLoggerConfig` and I added comments to explain that.

This is consistent with other languages in dealing with things like LB policies. We will only call this `Build` with the config returned from `ParseAuditLoggerConfig` and implementers must ensure a valid config from there is a valid one for building a logger.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:68]: Added more explanation + link to the ongoing gRFC. PTAL.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:80]: Same as above.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:28]: I am not aware of any convention. This is for consistency within this repository. For example, the map in the `balancer` pkg is named `m`.

Comment (gtcooke94) [reply authz/audit_logger.go:28]: Personally I'd be a fan of giving it at least some kind of meaningful name unless that is a very strong convention

Comment (rockspore) [reply authz/audit_logger.go]: I added a comment on this var following the `balancer` pkg. I still think consistency is more beneficial in that if someone needed to do something similar in the future, they wouldn't need to spend extra time thinking which prior art to follow. Let me know if you don't think this is good enough. ;)

File: authz/audit_logger.go
Comment (dfawley) [reply authz/audit_logger.go:99]: I see what you're saying about nacking the config, but there are failure modes caused by runtime problems even with a valid config.  Is the implication that if there is an error like this, the logger needs to keep retrying to recover?  Or should it go into a lame mode and just do nothing forever?  Is there some guidance we want to provide for potential users here, and should it go into the gRFC/a user guide/docstrings (in all languages)?

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: This being in the `authz` pkg is a strong point that I didn't realize. I changed the name to `auditLoggerBuilderRegistry`.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:48]: Is this JSON stuff here because we need it internally in order to make any audit logger work, or because we're planning to have one specific implementation of an audit logger that wants to output the data as JSON?  If the latter, then I don't think we want this here.

File: authz/audit_logger.go
Comment (dfawley) [reply authz/audit_logger.go]: Well now it feels like Java...

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:29]: Remove this line? It also backs the GetAuditLoggerBuilder API, and might be used for other things later.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:107]: Nit: let's move this up next to the other registry-related functions.

File: authz/audit_logger.go
Comment (dfawley) [reply authz/audit_logger.go:64]: Delete `Name()` now?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:79]: Should this return an error?  E.g. what happens if the context expires while it's executing?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:76]: It's unfortunate this is still pending.  How will we remember to update this when it's merged?  Can we wait until then to merge this?

Comment (rockspore) [reply authz/audit_logger.go:83]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:100]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:112]: Per our offline discussion, I think we agreed that the builder should never fail to create a logger given a validated config. Runtime issues should be handled by the logger impl. I documented this in the gRFC.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go:83]: "Done" ^ was for the blank `//`. I added a `TODO` to redirect the link to a merged gRFC. I apologize for accumulating tech debt so early but it'd be nice to unblock some of the implementations (IIUC, we need these interfaces when adding the parsing logic in http rbac filter for the logger config, as well as invoking loggers in rbac engines) before the gRFC is completely finalized (as least [one thing](https://github.com/envoyproxy/envoy/pull/26453) outside gRPC that's blocking it).

But of course we should definitely fix it in this PR for anything still unclear with the APIs themselves.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Resolved with offline discussion. Removed the `context.Context` and no error shall be returned.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Removed.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Good point. I removed these annotations.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: I renamed it to `loggerBuilderRegistry`. Hopefully it's less like Java now ;-)

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:27]: This should be named in a way that indicates it protects the registry, since the scope of this package is more than the audit loggers.  Or grouped with it in a struct, either way.

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:55]: How fixed are these names?  We've called this field `FullMethodName` in the past to make it more obvious it includes the service name + method name.  (See: balancer, stats, and tap packages)

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:51]: Bikeshedding: hmm, maybe `AuditEvent` is more precise?

File: authz/audit_logger.go
Comment (dfawley) [inline authz/audit_logger.go:86]: s/should/must/ please

**Comment by dfawley** [comment]
FYI I'll be OOO next week, but @easwars can approve in my absence.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Makes sense! I grouped it in a struct.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: True. Changed to `AuditEvent`.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

**Comment by rockspore** [comment]
> FYI I'll be OOO next week, but @easwars can approve in my absence.

Could you PTAL when you got time, @easwars? Thank you!

File: authz/audit_logger.go
Comment (easwars) [reply authz/audit_logger.go]: Was a separate package for audit logging inside the `authz` directory considered? If so, what were the reasons for not taking that approach? I feel that it might be cleaner to have an `auditlog` package inside the `authz` directory, since audit logger is a reasonably separate component in itself. 

Also, I don't see the audit logging functionality depend on anything in the `authz` package. So, I believe circular dependencies might not be an issue here.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: It's the identity of the caller, extracted at TLS connection layer. I changed the description.

**Comment by rockspore** [comment]
Regarding a separate `auditlog` pkg, that was the original proposal and was discussed offline. While I am pretty neutral on this, IIRC, the reasoning was that separate packages always added incremental costs and everything here would only be consumed within `authz` with no plan to add anything to make audit logging more standalone and separate from authorization. Therefore, although variable and func names have to be more verbose, putting everything in `authz` actually makes it clear and convenient for users.

File: authz/audit_logger.go
Comment (easwars) [inline authz/audit_logger.go:101]: How about:

```suggestion
	// Log performs audit logging for the provided audit event.
	//
	// This method is invoked in the RPC path and therefore implementations
	// must not block.
```

File: authz/audit_logger.go
Comment (easwars) [inline authz/audit_logger.go:92]:
```suggestion
// AuditLogger is the interface to be implemented by audit loggers.
//
// An audit logger is a logger instance that can be configured via the
// authorization policy API or xDS HTTP RBAC filters. When the authorization
// decision meets the condition for audit, all the configured audit loggers'
// Log() method will be invoked to log that event.
```

Skipping the line about the method being invoked inline since the same is mentioned in the docstring for `Log`.

File: authz/audit_logger.go
Comment (easwars) [inline authz/audit_logger.go:112]:
```suggestion
// AuditLoggerBuilder is the interface to be implemented by audit logger
// builders that are used at runtime to configure and instantiate audit loggers.
//
// Users who want to implement their own audit logging logic should
// implement this interface, along with the AuditLogger interface, and register it
// by calling RegisterAuditLoggerBuilder() at init time.
//
```

File: authz/audit_logger.go
Comment (easwars) [inline authz/audit_logger.go:120]: This is mentioned in the docstring of the `AuditLoggerConfig`. I would consider skipping it here.

**Review by easwars — CHANGES_REQUESTED** [review]

**Comment by rockspore** [comment]
Thanks a lot for the suggested revision of those comments! They definitely document the APIs in a clearer way. I took all the suggestions. @easwars

File: authz/audit_logger.go
Comment (rockspore) [reply authz/audit_logger.go]: Done.

**Review by gtcooke94 — APPROVED** [review]
LGTM

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
@rockspore 
How soon do you want this merged? I see that @dfawley had requested some changes and I think you have handled all of those. If this can wait till next week, then we can wait for him to come back, take another look and then merge. If this needs to go in asap, I will have to dismiss his review to be able to merge this. Let me know how you want to proceed. Thanks.

**Comment by rockspore** [comment]
> @rockspore How soon do you want this merged? I see that @dfawley had requested some changes and I think you have handled all of those. If this can wait till next week, then we can wait for him to come back, take another look and then merge. If this needs to go in asap, I will have to dismiss his review to be able to merge this. Let me know how you want to proceed. Thanks.

Thanks. I'll leave this to @gtcooke94 who will follow up with the rest of work in Go.

**Review by dfawley — APPROVED** [review]

**Pull request merged**
